pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,511 @@
+package index
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/common/model"
+	"golang.org/x/sync/errgroup"
+
+	metastorev1 "github.com/grafana/pyroscope/api/gen/proto/go/metastore/v1"
+)
+
+// NewIndex initializes a new metastore index.
+//
+// The index provides access to block metadata. The data is partitioned by time, shard and tenant. Partition identifiers
+// contain the time period referenced by partitions, e.g., "20240923T16.1h" refers to a partition for the 1-hour period
+// between 2024-09-23T16:00:00.000Z and 2024-09-23T16:59:59.999Z.
+//
+// Partitions are mostly transparent for the end user, though PartitionMeta is at times used externally. Partition
+// durations are configurable (at application level).
+//
+// The index requires a backing Store for loading data in memory. Data is loaded directly via LoadPartitions() or when
+// looking up blocks with FindBlock() or FindBlocksInRange().
+//
+// Data can be unloaded via StartCleanupLoop() which is meant to be called by the index owner and ran in a goroutine.
+func NewIndex(store Store, logger log.Logger, cfg *Config) *Index {
+	return &Index{
+		loadedPartitions: make(map[PartitionKey]*indexPartition),
+		allPartitions:    make([]*PartitionMeta, 0),
+		store:            store,
+		logger:           logger,
+		Config:           cfg,
+	}
+}
+
+// LoadPartitions reads all partitions from the backing store and loads the recent ones in memory.
+func (i *Index) LoadPartitions() {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	i.loadedPartitions = make(map[PartitionKey]*indexPartition)
+	i.allPartitions = make([]*PartitionMeta, 0)
+	for _, key := range i.store.ListPartitions() {
+		pMeta, err := i.store.ReadPartitionMeta(key)
+		if err != nil {
+			level.Error(i.logger).Log("msg", "error reading partition metadata", "key", key, "err", err)
+			continue
+		}
+		i.allPartitions = append(i.allPartitions, pMeta)
+		if pMeta.Ts.Add(i.Config.PartitionTTL).Before(time.Now()) {
+			// too old, will load on demand
+			continue
+		}
+		_, _ = i.getOrLoadPartition(pMeta)
+	}
+
+	i.sortPartitions()
+}
+
+// ForEachPartition executes the given function concurrently for each partition. It will be called for all partitions,
+// regardless if they are fully loaded in memory or not.
+func (i *Index) ForEachPartition(ctx context.Context, fn func(meta *PartitionMeta)) {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	g, ctx := errgroup.WithContext(ctx)
+	for _, meta := range i.allPartitions {
+		g.Go(func() error {
+			fn(meta)
+			return nil
+		})
+	}
+	err := g.Wait()
+	if err != nil {
+		level.Error(i.logger).Log("msg", "error during partition iteration", "err", err)
+	}
```

Comment:
Shouldn't we return the error? I think it's up to the caller to decide how to handle it